### PR TITLE
V22F-699 Environment Selection

### PR DESCRIPTION
**Pull request by Sindarin27**
Allow selecting environments from the Researcher Menu. Use the selected environment when starting a race.

### CODE REVIEW

File: speed-dreams/source-2.2.3/src/modules/userinterface/textonly/textonly.cpp [hunk 1]
```diff
@@ -116,13 +117,15 @@ bool LoadReplayConfiguration(GfRaceManager*& p_selRaceMan) {
...
     const char* trackCategory = GfParmGetStr(replaySettingsHandle, PATH_TRACK, KEY_CATEGORY, nullptr);
     const char* trackName = GfParmGetStr(replaySettingsHandle, PATH_TRACK, KEY_NAME, nullptr);
+    // Recreate the path of the environment descriptor file, which is always at "tracks/[category]/[name]/[name].xml"
+    std::stringstream trackFilename("tracks/");
+    trackFilename << trackCategory << "/" << trackName << "/" << trackName << ".xml";
+    SMediator::GetInstance()->SetEnvironmentFilePath(trackFilename.str().c_str());
+
 
     p_selRaceMan = GfRaceManagers::self()->getRaceManager("replay");
     auto handle = p_selRaceMan->getDescriptorHandle();
```
Comment (suppergerrie2) [inline speed-dreams/source-2.2.3/src/modules/userinterface/textonly/textonly.cpp:127]: Unused now

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/ConfigTests.cpp [hunk 2]
```diff
@@ -154,6 +154,34 @@ TEST(ConfigTests, BlackBoxFilePathTest)
...
+        for (int i = 0; i < length - 1; i++)
+        {
+            char c;
+            do
+            {  // Make sure no character is assigned the null character
+                c = (char)random.NextByte();
+            } while (c == '\0');
+            path[i] = c;
+        }
+        path[length - 1] = '\0';
```
Comment (suppergerrie2) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/ConfigTests.cpp:175]: Probably good to use the RandomCharArray function

**Review by suppergerrie2 — CHANGES_REQUESTED** [review]

**Review by larsdekwant — CHANGES_REQUESTED** [review]
- Evenly spread out controls on the researchermenu (currently there is a large gap between Intervention Type and Participant Control)
- When you go to racestart menu by clicking on Start on the mainmenu, the background is determined by what is selected in the 'Configure' menu of that screen, and not what you selected in the researchermenu. I think the whole Configure button should be removed, since the trackselection that is contained within that menu is completely ignored (but still changes the background of the racestartmenu)

**Comment by Sindarin27** [comment]
Removal of Configure button will be done in a separate issue.

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/racescreens/raceconfigstate.cpp [hunk 3]
```diff
@@ -147,8 +148,9 @@ RmConfigRunState(bool bStart)
 		} else {
 			ts.prevScreen = rmConfigBackHookInit();
 		}
-		ts.pRace = LmRaceEngine().race();
-		ts.piTrackLoader = GfTracks::self()->getTrackLoader();
+                ts.SetTrack = [](GfTrack* p_track){LmRaceEngine().race()->getManager()->setEventTrack(0, p_track);};
+                ts.GetTrack = []{return LmRaceEngine().race()->getTrack();};
+                ts.piTrackLoader = GfTracks::self()->getTrackLoader();
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/racescreens/raceconfigstate.cpp:153]: // SIMULATED DRIVING ASSISTANCE

File: speed-dreams/source-2.2.3/data/data/menu/ResearcherMenu.xml [hunk 4]
```diff
@@ -347,22 +353,13 @@
...
-        <section name="5">
-            <attstr name="type" val="label"/>
-            <attnum name="x" val="380"/>
-            <attnum name="y" val="400"/>
-            <attstr name="text" val="Environment"/>
-            <attstr name="h align" val="left"/>
-            <attstr name="font" val="medium_t"/>
-        </section>
-
         <section name="6">
```
Comment (suppergerrie2) [inline speed-dreams/source-2.2.3/data/data/menu/ResearcherMenu.xml:363]: Shouldn't this be 5 now?

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp [hunk 5]
```diff
@@ -254,6 +302,14 @@ static void SaveSettingsToDisk()
     // Save black box filepath to xml file
     GfParmSetStr(readParam, PRM_BLACKBOX, GFMNU_ATTR_PATH, m_blackBoxFilePath);
 
+    // Save environment filepath to xml file if a track has been selected
+    // Speed Dreams identifies tracks by their category and their name, so save these two
+    if (m_environmentChosen)
+    {
+        GfParmSetStr(readParam, PRM_ENVIRONMENT_CATEGORY, GFMNU_ATTR_PATH, m_environment->getCategoryId().c_str());
+        GfParmSetStr(readParam, PRM_ENVIRONMENT_NAME, GFMNU_ATTR_PATH, m_environment->getId().c_str());
```
Comment (suppergerrie2) [inline speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp:310]: Why not put them in `Environment` category with `category` and `name` as keys:

```xml
  <section name="Environment">
    <attstr name="category" val="road"/>
    <attstr name="name" val="test-highway"/>
  </section>
  ```
  
  VS
  
  ```xml
  

  <section name="EnvironmentCategory">
    <attstr name="path" val="road"/>
  </section>

  <section name="EnvironmentName">
    <attstr name="path" val="test_highway"/>
  </section>
  ```

**Review by suppergerrie2 — CHANGES_REQUESTED** [review]

**Review by suppergerrie2 — APPROVED** [review]

**Review by larsdekwant — APPROVED** [review]

**Pull request merged**